core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RebalanceOnlyWhenActiveAllocationDecider.java

```diff
@@ -40,7 +40,8 @@ public RebalanceOnlyWhenActiveAllocationDecider(Settings settings) {
     public Decision canRebalance(ShardRouting shardRouting, RoutingAllocation allocation) {
         // its ok to check for active here, since in relocation, a shard is split into two in routing
         // nodes, once relocating, and one initializing
-        if (!allocation.routingNodes().allReplicasActive(shardRouting)) {
+        int expectedAssignedShardsCount = allocation.routingTable().index(shardRouting.index().getName()).shard(shardRouting.id()).size();
```

Comment:
not sure if that is an improvement to be honest? it seems like it's way more work to get this outside of the method? maybe have an overloaded version?